app/models/form_item.rb

```diff
@@ -27,8 +31,26 @@ def arrange_and_sort(eager_load = nil)
       order: '(case when ancestry is null then 0 else 1 end), ancestry, rank'))
   end
 
+  # tests for cyclic parents
+  def check_ancestry_integrity(parent_id)
+    parent = FormItem.find_by_id(parent_id)
+    return true if parent.nil?
+    parent.ancestry != self.id.to_s
+  end
+
   private
 
+    # sets rank if not already set
+    def set_rank
+      self.rank ||= (form.try(:max_rank) || 0) + 1
+      return true
+    end
```

Comment:
This method is not aware of the new hierarchical system. It should not be using form.max_rank since that returns the max rank for the topmost level questions only. Rather it should be looking on its own level. In general the old form rank methods are not to be trusted as they are not really aware of the hierarchical stuff.